File: windows/daemon/Networking/PeerConnection.cs
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using ClipboardDaemon.Identity;
using ClipboardDaemon.Crypto;

namespace ClipboardDaemon.Networking;

public class PeerConnection
{
    // Sent (encrypted, like anything else on this channel) as a heartbeat
    // and filtered out in Listen() before reaching MessageReceived - normal
    // message handling never sees these. TCP keepalive alone (see
    // Program.cs's EnableKeepAlive) isn't enough by itself to catch a
    // silently dead peer in reasonable time - depending on OS/network
    // conditions it can still take much longer than feels "live". This
    // application-level ping/watchdog is what actually makes a connection
    // list reflect reality within seconds of a peer actually going away.
    private const string PingSentinel = "__ping__";
    private const int HeartbeatIntervalMs = 5000;
    private const int HeartbeatTimeoutMs = 15000;

    private readonly TcpClient client;
    private readonly StreamReader reader;
    private readonly StreamWriter writer;
    private readonly byte[] sessionKey;
    private DateTime lastActivityAt = DateTime.UtcNow;
    private System.Threading.Timer? heartbeatTimer;
    // StreamWriter isn't safe for concurrent calls - without this, the
    // heartbeat timer's own Send(ping) can race a real Send() from, say,
    // StreamFileToPeer's chunk loop (different threads entirely) and
    // corrupt the writer's internal state. That's what an
    // ArgumentOutOfRangeException out of WriteLineAsync actually was -
    // this serializes every Send() so only one write is ever in flight.
    private readonly SemaphoreSlim writeLock = new(1, 1);

    // Known the moment the connection is created — verified during the
    // handshake, not inferred later from message content.
    public string PeerDeviceId { get; }

    // False means this peer wasn't in the trust store when the handshake
    // ran, and the connection only exists at all because pairingModeOpen
    // was true (see CreateAsync's relaxed gate below). Callers MUST treat
    // this as a live pairing candidate — show an explicit accept/reject
    // prompt fed by this already signature-verified PeerDeviceId — rather
    // than handing it straight to normal message processing.
    public bool WasAlreadyTrusted { get; }

    // True only when WasAlreadyTrusted became true because THIS handshake's
    // passphrase proof verified (not because the peer was already in the
    // trust store) - the caller's signal to actually WRITE trust for this
    // peer before treating the connection as normal, same as the LAN
    // beacon's own auto-trust path already does.
    public bool NewlyTrustedViaPassphrase { get; }

    public event Action<string>? MessageReceived;
    public event Action? Disconnected;

    private PeerConnection(TcpClient client, StreamReader reader, StreamWriter writer, byte[] sessionKey, string peerDeviceId, bool wasAlreadyTrusted, bool newlyTrustedViaPassphrase)
    {
        this.client = client;
        this.reader = reader;
        this.writer = writer;
        this.sessionKey = sessionKey;
        PeerDeviceId = peerDeviceId;
        WasAlreadyTrusted = wasAlreadyTrusted;
        NewlyTrustedViaPassphrase = newlyTrustedViaPassphrase;
    }

    // Performs the authenticated ECDH handshake described in HandshakeMessage.cs.
    // Returns null (never throws for a bad peer, or an untrusted one that's
    // neither pairing-mode-eligible nor passphrase-verified) if the
    // handshake fails or the signature doesn't verify — callers should just
    // close the socket and move on.
    //
    // pairingModeOpen: true only while the local tray's pairing dialog is
    // open (mirrors HarmonyOS's Index.ets pairingOpen) — a live, explicit
    // "I'm expecting to pair right now" signal. Without it, an untrusted
    // peer is refused exactly as before (same early-bail, no wasted
    // signature verification) UNLESS their handshake carries a passphrase
    // proof that verifies against passphraseKeyStore (see
    // NewlyTrustedViaPassphrase) - a second, independent way in, same as
    // the LAN beacon's passive auto-trust already is regardless of pairing
    // mode. With pairingModeOpen true, the handshake completes fully for a
    // live accept/reject prompt instead - both sides need this true on
    // their OWN behalf for a NEW pairing to succeed that way, since each
    // independently runs this same check over the same connection; the
    // caller who gets back WasAlreadyTrusted=false is responsible for a
    // local accept/reject prompt before writing trust.
    public static async Task<PeerConnection?> CreateAsync(TcpClient client, DeviceIdentity myIdentity, TrustStore trustStore, bool pairingModeOpen, PassphraseKeyStore passphraseKeyStore)
    {
        Stream stream = client.GetStream();
        var reader = new StreamReader(stream);
        var writer = new StreamWriter(stream) { AutoFlush = true };

        using var ecdh = ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256);
        byte[] myEphemeralPublicKey = ecdh.PublicKey.ExportSubjectPublicKeyInfo();
        byte[] mySignature = myIdentity.SignData(myEphemeralPublicKey);

        byte[]? myKey = passphraseKeyStore.GetKey();
        string? myProof = myKey != null ? PassphraseAuth.ComputeProof(myKey, myIdentity.GetPublicKey()) : null;

        var myHandshake = new HandshakeMessage(
            Convert.ToBase64String(myEphemeralPublicKey),
            myIdentity.GetPublicKey(),
            Convert.ToBase64String(mySignature),
            myProof);
        await writer.WriteLineAsync(JsonSerializer.Serialize(myHandshake));

        string? theirHandshakeJson = await reader.ReadLineAsync();
        if (theirHandshakeJson == null) return null;

        HandshakeMessage? theirHandshake;
        try
        {
            theirHandshake = JsonSerializer.Deserialize<HandshakeMessage>(theirHandshakeJson);
        }
        catch (JsonException) { return null; }
        if (theirHandshake == null) return null;

        bool alreadyTrusted = trustStore.IsTrusted(theirHandshake.IdentityPublicKey);
        bool passphraseVerified = false;
        if (!alreadyTrusted && myKey != null && theirHandshake.PassphraseProof != null)
        {
            passphraseVerified = PassphraseAuth.VerifyProof(myKey, theirHandshake.IdentityPublicKey, theirHandshake.PassphraseProof);
        }
        bool effectivelyTrusted = alreadyTrusted || passphraseVerified;
        if (!effectivelyTrusted && !pairingModeOpen)
        {
            return null; // not a device we trust or can auto-trust, and not expecting to pair right now — refuse outright
        }

        byte[] theirEphemeralPublicKeyBytes;
        byte[] theirSignature;
        try
        {
            theirEphemeralPublicKeyBytes = Convert.FromBase64String(theirHandshake.EphemeralPublicKey);
            theirSignature = Convert.FromBase64String(theirHandshake.Signature);
        }
        catch (FormatException) { return null; }

        using var verifyEcdsa = ECDsa.Create();
        try
        {
            verifyEcdsa.ImportSubjectPublicKeyInfo(Convert.FromBase64String(theirHandshake.IdentityPublicKey), out _);
        }
        catch (CryptographicException) { return null; }

        if (!verifyEcdsa.VerifyData(theirEphemeralPublicKeyBytes, theirSignature, HashAlgorithmName.SHA256))
        {
            return null; // ephemeral key doesn't match the claimed identity's signature — tampered or spoofed
        }

        using var theirEcdh = ECDiffieHellman.Create();
        theirEcdh.ImportSubjectPublicKeyInfo(theirEphemeralPublicKeyBytes, out _);

        byte[] sessionKey = ecdh.DeriveKeyFromHash(theirEcdh.PublicKey, HashAlgorithmName.SHA256);

        return new PeerConnection(client, reader, writer, sessionKey, theirHandshake.IdentityPublicKey, effectivelyTrusted, passphraseVerified);
    }

    public async Task Send(string message)
    {
        string encrypted = Encrypt(message);
        await writeLock.WaitAsync();
        try
        {
            await writer.WriteLineAsync(encrypted);
        }
        finally
        {
            writeLock.Release();
        }
    }

    public async Task Listen()
    {
        StartHeartbeat();
        try
        {
            while (true)
            {
                string? line = await reader.ReadLineAsync();
                if (line == null) break; // peer closed cleanly
                lastActivityAt = DateTime.UtcNow;
                string decrypted = Decrypt(line);
                if (decrypted == PingSentinel) continue; // heartbeat only, not real data
                MessageReceived?.Invoke(decrypted);
            }
        }
        catch (Exception) { } // abrupt disconnect, or a corrupt/forged line that failed to decrypt
        finally
        {
            StopHeartbeat();
            Disconnected?.Invoke();
        }
    }

    private void StartHeartbeat()
    {
        heartbeatTimer = new System.Threading.Timer(_ =>
        {
            _ = Send(PingSentinel); // fire-and-forget - a send failure also surfaces via Listen()'s read loop dying
            if ((DateTime.UtcNow - lastActivityAt).TotalMilliseconds > HeartbeatTimeoutMs)
            {
                // Nothing at all (ping or real data) for HeartbeatTimeoutMs -
                // the peer is gone even though TCP itself hasn't noticed
                // yet. Closing here is what actually makes the connections
                // list reflect reality, instead of relying on some much
                // longer OS-level timeout to eventually notice.
                Close();
            }
        }, null, HeartbeatIntervalMs, HeartbeatIntervalMs);
    }

    private void StopHeartbeat()
    {
        heartbeatTimer?.Dispose();
        heartbeatTimer = null;
    }

    // Actively tears down the connection — used when a device gets untrusted
    // while still connected.
    public void Close()
    {
        StopHeartbeat();
        client.Close();
    }

    private const int NonceSize = 12; // AES-GCM standard nonce size
    private const int TagSize = 16;

    private string Encrypt(string plaintext)
    {
        byte[] nonce = RandomNumberGenerator.GetBytes(NonceSize);
        byte[] plainBytes = Encoding.UTF8.GetBytes(plaintext);
        byte[] cipherBytes = new byte[plainBytes.Length];
        byte[] tag = new byte[TagSize];

        using var aesGcm = new AesGcm(sessionKey, TagSize);
        aesGcm.Encrypt(nonce, plainBytes, cipherBytes, tag);

        byte[] packed = new byte[NonceSize + TagSize + cipherBytes.Length];
        Buffer.BlockCopy(nonce, 0, packed, 0, NonceSize);
        Buffer.BlockCopy(tag, 0, packed, NonceSize, TagSize);
        Buffer.BlockCopy(cipherBytes, 0, packed, NonceSize + TagSize, cipherBytes.Length);
        return Convert.ToBase64String(packed);
    }

    private string Decrypt(string packedBase64)
    {
        byte[] packed = Convert.FromBase64String(packedBase64);
        byte[] nonce = packed[0..NonceSize];
        byte[] tag = packed[NonceSize..(NonceSize + TagSize)];
        byte[] cipherBytes = packed[(NonceSize + TagSize)..];
        byte[] plainBytes = new byte[cipherBytes.Length];

        using var aesGcm = new AesGcm(sessionKey, TagSize);
        aesGcm.Decrypt(nonce, cipherBytes, tag, plainBytes); // throws if tampered — caught in Listen()
        return Encoding.UTF8.GetString(plainBytes);
    }
}

File: mobile/android/ClipLink/app/src/main/java/io/uaena/cliplink/net/PeerConnection.kt
package io.uaena.cliplink.net

import io.uaena.cliplink.core.AesGcm
import io.uaena.cliplink.core.B64
import io.uaena.cliplink.core.DeviceIdentity
import io.uaena.cliplink.store.PassphraseKeyStore
import io.uaena.cliplink.store.TrustStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.Socket
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.spec.ECGenParameterSpec
import java.security.spec.X509EncodedKeySpec
import javax.crypto.KeyAgreement
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/**
 * One authenticated, encrypted link to a peer.
 *
 * The handshake is symmetric: both sides write their own handshake line
 * first, then read the other's. There is no client/server role, so the same
 * code path runs whether this device dialled out or was dialled.
 *
 * Framing is newline-delimited UTF-8 for the whole connection's life,
 * handshake and session alike. The Windows daemon writes `\r\n` and the other
 * two write `\n`; every reader tolerates both.
 */
class PeerConnection private constructor(
    private val socket: Socket,
    private val reader: BufferedReader,
    private val writer: BufferedWriter,
    private val sessionKey: ByteArray,
    val peerDeviceId: String,
    /**
     * False means this peer was not in the trust store when the handshake
     * ran, and the connection exists only because pairing mode was open. The
     * caller MUST surface an explicit accept/reject prompt rather than hand
     * it to the sync manager like an ordinary trusted link.
     */
    val wasAlreadyTrusted: Boolean,
    /**
     * True only when trust was established by THIS handshake's passphrase
     * proof verifying - the caller's signal to actually persist trust.
     */
    val newlyTrustedViaPassphrase: Boolean,
) {

    private val writeLock = Mutex()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val lastActivityAt = AtomicLong(System.currentTimeMillis())
    private val closed = AtomicBoolean(false)
    private val disconnectFired = AtomicBoolean(false)

    var onMessage: ((String) -> Unit)? = null
    var onDisconnected: (() -> Unit)? = null

    val remoteAddress: String? get() = socket.inetAddress?.hostAddress

    /**
     * A connection can die between the handshake returning and the caller
     * registering it - at which point onDisconnected was still null and will
     * never fire, leaving a dead link registered as live forever. Callers
     * check this after wiring their callbacks.
     */
    val isClosed: Boolean get() = closed.get()

    suspend fun send(message: String) = withContext(Dispatchers.IO) {
        val encrypted = AesGcm.encrypt(sessionKey, message)
        writeLine(encrypted)
    }

    /**
     * Serialised because the heartbeat writes on its own schedule and can
     * otherwise interleave with a real message mid-stream. The Windows side
     * carries the identical lock for the identical reason.
     */
    private suspend fun writeLine(line: String) = writeLock.withLock {
        writer.write(line)
        writer.write("\n")
        writer.flush()
    }

    /** Starts the read loop and the heartbeat. A connection isn't live for either until this. */
    fun listen() {
        scope.launch { readLoop() }
        scope.launch { heartbeatLoop() }
    }

    private suspend fun readLoop() = withContext(Dispatchers.IO) {
        try {
            while (true) {
                val line = reader.readLine() ?: break
                if (line.isEmpty()) continue
                val decrypted = AesGcm.decrypt(sessionKey, line)
                lastActivityAt.set(System.currentTimeMillis())
                if (decrypted == PING_SENTINEL) continue // heartbeat, never real data
                onMessage?.invoke(decrypted)
            }
        } catch (e: Exception) {
            // Clean close, abrupt disconnect, or a corrupt/forged line that
            // failed to decrypt. All three end the connection, same as the
            // single catch-all on the other two platforms.
        } finally {
            finish()
        }
    }

    private suspend fun heartbeatLoop() {
        while (scope.coroutineContext.isActive) {
            delay(HEARTBEAT_INTERVAL_MS)
            if (closed.get()) return
            try {
                send(PING_SENTINEL)
            } catch (e: Exception) {
                // A dead write also surfaces via the read loop ending.
            }
            if (System.currentTimeMillis() - lastActivityAt.get() > HEARTBEAT_TIMEOUT_MS) {
                // Nothing at all for the timeout window: the peer is gone even
                // though TCP hasn't noticed. Closing here is what actually makes
                // the UI's connected count reflect reality in seconds rather than
                // whenever the OS eventually gives up, which can be hours -
                // Android exposes no per-socket keepalive interval either.
                close()
                return
            }
        }
    }

    fun close() = finish()

    /**
     * Idempotent and safe from either side of the race - the read loop's
     * `finally` and an explicit [close] routinely both land here. Firing
     * onDisconnected twice would make the sync manager evict a connection it
     * has already replaced, which is precisely the reconnect loop this
     * protocol had to be fixed for once already.
     */
    private fun finish() {
        if (!closed.getAndSet(true)) {
            try {
                socket.close()
            } catch (e: Exception) {
                // already closed - fine
            }
        }
        if (disconnectFired.getAndSet(true)) return
        scope.coroutineContext[Job]?.cancel()
        onDisconnected?.invoke()
    }

    companion object {
        private const val PING_SENTINEL = "__ping__"

        /**
         * Deliberately tighter than the Windows daemon's 5000/15000. What has
         * to hold is that EACH side pings faster than the OTHER side's
         * timeout, and both directions satisfy that: this pings every 3s
         * against the daemon's 15s timeout, the daemon pings every 5s against
         * this 9s timeout.
         */
        private const val HEARTBEAT_INTERVAL_MS = 3000L
        private const val HEARTBEAT_TIMEOUT_MS = 9000L

        private const val HANDSHAKE_TIMEOUT_MS = 10_000

        /**
         * Runs the handshake and returns a live connection, or null for any
         * failure at all - a malformed peer, a bad signature, or an untrusted
         * one that is neither pairing-eligible nor passphrase-verified. Never
         * throws, matching CreateAsync's contract on the Windows side.
         *
         * [pairingModeOpen] must be a live "I am at the pairing screen right
         * now" signal, not a setting. With it false, an untrusted peer is
         * refused before this even spends effort verifying their signature.
         * With it true, the handshake completes so the caller can show an
         * accept/reject prompt. A verifying passphrase proof is a second,
         * independent way in, exactly as the LAN beacon's passive auto-trust
         * already is.
         */
        suspend fun create(
            socket: Socket,
            identity: DeviceIdentity,
            trustStore: TrustStore,
            passphraseKeyStore: PassphraseKeyStore,
            pairingModeOpen: Boolean,
        ): PeerConnection? = withContext(Dispatchers.IO) {
            val connection = try {
                handshake(socket, identity, trustStore, passphraseKeyStore, pairingModeOpen)
            } catch (e: Exception) {
                null
            }
            if (connection == null) {
                // Every failure path closes the socket here rather than
                // leaving it to each caller - a half-open socket per refused
                // handshake adds up fast when a stranger's beacon arrives
                // every two seconds.
                try {
                    socket.close()
                } catch (ignored: Exception) {
                    // already closed - fine
                }
            }
            connection
        }

        private fun handshake(
            socket: Socket,
            identity: DeviceIdentity,
            trustStore: TrustStore,
            passphraseKeyStore: PassphraseKeyStore,
            pairingModeOpen: Boolean,
        ): PeerConnection? {
            socket.soTimeout = HANDSHAKE_TIMEOUT_MS
            socket.tcpNoDelay = true
            val reader = BufferedReader(InputStreamReader(socket.getInputStream(), Charsets.UTF_8))
            val writer = BufferedWriter(OutputStreamWriter(socket.getOutputStream(), Charsets.UTF_8))

            val ephemeral = KeyPairGenerator.getInstance("EC").run {
                initialize(ECGenParameterSpec("secp256r1"))
                generateKeyPair()
            }
            val ephemeralPublicBytes = ephemeral.public.encoded
            val myIdentityPublicKey = identity.publicKeyBase64()
            val myKey = passphraseKeyStore.key()

            val mine = HandshakeMessage(
                ephemeralPublicKey = B64.encode(ephemeralPublicBytes),
                identityPublicKey = myIdentityPublicKey,
                signature = B64.encode(identity.sign(ephemeralPublicBytes)),
                passphraseProof = myKey?.let {
                    passphraseKeyStore.computeProof(it, myIdentityPublicKey)
                },
            )
            writer.write(mine.toJson())
            writer.write("\n")
            writer.flush()

            val theirs = HandshakeMessage.parse(reader.readLine() ?: return null) ?: return null

            val alreadyTrusted = trustStore.isTrusted(theirs.identityPublicKey)
            val passphraseVerified = !alreadyTrusted && myKey != null &&
                passphraseKeyStore.verifyProof(
                    myKey,
                    theirs.identityPublicKey,
                    theirs.passphraseProof,
                )
            val effectivelyTrusted = alreadyTrusted || passphraseVerified
            if (!effectivelyTrusted && !pairingModeOpen) {
                // Not someone we trust or can auto-trust, and we aren't
                // expecting to pair - refuse before doing any more work.
                return null
            }

            val theirEphemeralBytes = B64.decodeOrNull(theirs.ephemeralPublicKey) ?: return null
            val theirSignature = B64.decodeOrNull(theirs.signature) ?: return null
            if (!DeviceIdentity.verifyRawSignature(
                    theirs.identityPublicKey,
                    theirEphemeralBytes,
                    theirSignature,
                )
            ) {
                return null
            }

            val theirEphemeralPublic = KeyFactory.getInstance("EC")
                .generatePublic(X509EncodedKeySpec(theirEphemeralBytes))
            val agreement = KeyAgreement.getInstance("ECDH").apply {
                init(ephemeral.private)
                doPhase(theirEphemeralPublic, true)
            }
            // Matches .NET's DeriveKeyFromHash(theirPublicKey, SHA256) with no
            // prepend/append - that is plain SHA256(rawSharedSecret), NOT
            // HKDF. Adding a salt or info here would derive a different key
            // and every message on the session would fail to decrypt.
            val sessionKey = MessageDigest.getInstance("SHA-256")
                .digest(agreement.generateSecret())

            // The handshake timeout must not outlive the handshake: the
            // session read loop parks in readLine() indefinitely by design,
            // and leaving a 10s SO_TIMEOUT on would tear down a perfectly
            // healthy idle connection every 10 seconds.
            socket.soTimeout = 0

            return PeerConnection(
                socket = socket,
                reader = reader,
                writer = writer,
                sessionKey = sessionKey,
                peerDeviceId = theirs.identityPublicKey,
                wasAlreadyTrusted = effectivelyTrusted,
                newlyTrustedViaPassphrase = passphraseVerified,
            )
        }
    }
}
